chrometool/chrometool.go

```diff
@@ -138,3 +153,146 @@ func EvalNodeCenter(box *dom.BoxModel) (float64, float64, error) {
 
 	return x, y, nil
 }
+
+func SelectNthElement(ctx context.Context, sel NthChildSel) error {
+	var res interface{}
+	if err := chromedp.Evaluate(fmt.Sprintf(selectNthElementJS, sel.Selector, sel.Position), &res).Do(ctx); err != nil {
+		if err.Error() == "encountered an undefined value" {
+			err = nil
+		}
+		return err
+	}
+	return nil
+}
+
+func ClipboardCommand(ctx context.Context, command string) error {
+	var res interface{}
+	if err := chromedp.Evaluate(fmt.Sprintf(clipboardInteractJS, command), &res).Do(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// DispatchKeyEventToBrowser takes up the key's value to be dispatched to the browser usually the result of keybinding.InterpretKeycode
+// It clicks on the element if isFocussed is false and then sends the appropiate key with the modifiers
+// The modifiers here refer to the xproto modifiers not to be confused with the ones in input.DispatchKeyEvent
+func DispatchKeyEventToBrowser(ctx context.Context, sel NthChildSel, str string, modifiers uint16, isFocussed bool) error {
+	if isFocussed {
+		if err := chromedp.Run(ctx,
+			ClickNthElement(sel, chromedp.ByQueryAll),
+		); err != nil {
+			return err
+		}
+	}
+
+	ctx = cdp.WithExecutor(ctx, chromedp.FromContext(ctx).Target)
+	keyEvents, err := dispatchKeyEvent(ctx, sel, str, modifiers)
+	if err != nil {
+		return err
+	}
+	for _, k := range keyEvents {
+		if err := k.Do(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func dispatchKeyEvent(ctx context.Context, sel NthChildSel, str string, modifiers uint16) ([]*input.DispatchKeyEventParams, error) {
+	r := keyStrToRune[str]
+
+	// force \n -> \r
+	if r == '\n' {
+		r = '\r'
+	}
+
+	// if not known key, encode as unidentified
+	v, ok := kb.Keys[r]
+	if !ok {
+		return kb.EncodeUnidentified(r), nil
+	}
+
+	keyDown := input.DispatchKeyEventParams{
+		Key:                   v.Key,
+		Code:                  v.Code,
+		NativeVirtualKeyCode:  v.Native,
+		WindowsVirtualKeyCode: v.Windows,
+		Text:                  v.Text,
+		UnmodifiedText:        v.Unmodified,
+	}
+	if runtime.GOOS == "darwin" {
+		keyDown.NativeVirtualKeyCode = 0
+	}
+	if v.Shift {
+		keyDown.Modifiers |= input.ModifierShift
+	}
+	keyDown.Modifiers |= keyEventModifier(modifiers)
+
+	if keyDown.Modifiers&input.ModifierCommand > 0 && runtime.GOOS == "darwin" {
+		if err := clipboardAction(ctx, sel, str, keyDown.Modifiers); err != nil {
+			return nil, err
+		}
+		return nil, nil
+	}
+
+	keyUp := keyDown
+	keyDown.Type, keyUp.Type = input.KeyDown, input.KeyUp
+
+	return []*input.DispatchKeyEventParams{&keyDown, &keyUp}, nil
```

Comment:
Unnecessary pointers? pointers are nice when you want to modify stuff (or need it to be nil-able), but when not, pass by value and let golang decided whether to put stuff on stack or heap

Reply:
I use chromedps function, kb.EncodeUnidentified(r) for unidentified runes, that returns a *[]input.., so I used pointers here as well